test_unit.py

```diff
@@ -3570,6 +3571,262 @@ def test_local_state(self):
+                    [
+                        abi.Type.make_uint_type(16),
+                        abi.Type.make_tuple_type(
+                            [
+                                abi.Type.make_byte_type(),
+                                abi.Type.make_static_array_type(
+                                    abi.Type.make_address_type(), 10
+                                ),
+                            ]
+                        ),
+                    ]
+                ),
+                f"(uint16,(byte,address[10]))",
+            ),
+            (
+                abi.Type.make_tuple_type(
+                    [
+                        abi.Type.make_ufixed_type(256, 16),
+                        abi.Type.make_tuple_type(
+                            [
+                                abi.Type.make_tuple_type(
+                                    [
+                                        abi.Type.make_string_type(),
+                                    ]
+                                ),
+                                abi.Type.make_bool_type(),
+                                abi.Type.make_tuple_type(
+                                    [
+                                        abi.Type.make_address_type(),
+                                        abi.Type.make_uint_type(8),
+                                    ]
+                                ),
+                            ]
+                        ),
+                    ]
+                ),
+                f"(ufixed256x16,((string),bool,(address,uint8)))",
+            ),
+        ]
+        for test_case in test_cases:
+            self.assertEqual(str(test_case[0]), test_case[1])
+            self.assertEqual(
+                test_case[0], abi.Type.type_from_string(test_case[1])
+            )
+
+    def test_make_type_invalid(self):
+        # Test for invalid uint
+        invalid_type_sizes = (-1, 0, 9, 513, 1024)
+        for i in invalid_type_sizes:
+            with self.assertRaises(error.ABITypeError) as e:
+                abi.Type.make_uint_type(i)
+            self.assertIn(f"unsupported uint bitSize: {i}", str(e.exception))
+        with self.assertRaises(TypeError) as e:
+            abi.Type.make_uint_type()
+
+        # Test for invalid ufixed
+        invalid_precisions = (-1, 0, 161)
+        for i in invalid_type_sizes:
+            with self.assertRaises(error.ABITypeError) as e:
+                abi.Type.make_ufixed_type(i, 1)
+            self.assertIn(f"unsupported ufixed bitSize: {i}", str(e.exception))
+        for j in invalid_precisions:
+            with self.assertRaises(error.ABITypeError) as e:
+                abi.Type.make_ufixed_type(8, j)
+            self.assertIn(
+                f"unsupported ufixed precision: {j}", str(e.exception)
+            )
+
+    def test_type_from_string_invalid(self):
+        test_cases = (
+            # uint
+            # Whitespaces are allowed, e.g. "uint8 " or "uint 8" can be parsed
```

Comment:
I am not so sure if it is allowed, please refer to ARC#4.

Reply:
Ah yes, you are correct -- I left this as a note to myself but forgot about it. I think `isdecimal()` can be used to check whether the string literal is a valid integer size.